src/blk/kernel/KernelDevice.cc

```diff
@@ -541,6 +541,19 @@ void KernelDevice::_aio_stop()
   if (aio) {
     dout(10) << __func__ << dendl;
     aio_stop = true;
+    IOContext wakeup_ctx(cct, nullptr, false);
+    wakeup_ctx.num_running++;
+    std::list<aio_t> batch;
+    aio_t aio_op(&wakeup_ctx, fd_directs[WRITE_LIFE_NOT_SET]);
+
+    // IO_CMD_NOOP is not implemented in the Linux kernel; using IO_CMD_PREAD with a 0-byte read as a substitute.
+    if (use_ioring && ioring_queue_t::supported())
+       aio_op.iocb.aio_lio_opcode = IO_CMD_NOOP;
+    else
+       aio_op.iocb.aio_lio_opcode = IO_CMD_PREAD;
+    batch.push_back(aio_op);
+    int retries = 0;
+    io_queue->submit_batch(batch.begin(), batch.end(), &wakeup_ctx, &retries);
```

Comment:
Shouldn't we wait on wakeup_ctx before going forward with the shutdown?
And even more - IIUC by sending dummy op we're trying to make sure that all the previous ops are completed prior to shutdown, right?
If so are you sure that operation completions happen in absolutely the same order as they've been submitted? What if some reordering occurs in OS depths and dummy op returns faster than the actual operations?